### PR TITLE
Send mixpanel event when relanch headless

### PR DESCRIPTION
**Pull request by riemannulus**
SSIA

### CODE REVIEW

File: src/main/main.ts [hunk 1]
```diff
@@ -566,10 +571,11 @@ function initializeIpc() {
     event.returnValue = "Not supported platform.";
   });
 
-  ipcMain.on("mixpanel-track-event", async (_, eventName: string) => {
+  ipcMain.on("mixpanel-track-event", async (_, eventName: string, param: {}) => {
```
Comment (BasixKOR) [inline src/main/main.ts:574]:
```suggestion
  ipcMain.on("mixpanel-track-event", async (_, eventName: string, param: object) => {
```

File: src/main/main.ts [hunk 2]
```diff
@@ -538,7 +538,12 @@ function initializeIpc() {
     }
   });
 
-  ipcMain.on("relaunch standalone", async (event) => {
+  ipcMain.on("relaunch standalone", async (event, param: {}) => {
```
Comment (BasixKOR) [inline src/main/main.ts:541]: The type `{}`, confusingly, includes primitive types such as numbers and strings.

Personally I think it's better to use dedicated type like `Record<"reason", string>`, but either way looks okay to me.

```suggestion
  ipcMain.on("relaunch standalone", async (event, param: object) => {
```

File: src/renderer/views/preload/PreloadProgressView.tsx [hunk 3]
```diff
@@ -160,21 +160,21 @@ const PreloadProgressView = observer(() => {
...
           "Haven't received any messages for some time. Automatically relaunch."
         );
-        ipcRenderer.send("relaunch standalone");
+        ipcRenderer.send("relaunch standalone", {reason: "Haven't received message."});
         break;
       case 0x05:
         console.error("Action Timeout. Automatically relaunch.");
-        ipcRenderer.send("relaunch standalone");
+        ipcRenderer.send("relaunch standalone", {reason: "Action Timeout."});
         break;
```
Comment (BasixKOR) [inline src/renderer/views/preload/PreloadProgressView.tsx:178]: Seems like the fallback case is missing. Is it intended?

File: src/main/main.ts (hunk 2 above)
Comment (riemannulus) [reply src/main/main.ts:541]: I intend for give choice to caller. I guess `object` is looks good to me.

File: src/renderer/views/preload/PreloadProgressView.tsx (hunk 3 above)
Comment (riemannulus) [reply src/renderer/views/preload/PreloadProgressView.tsx:178]: IDK why `default` fallback is missing. but I'm just do not fix what I don't know.

Comment (BasixKOR) [reply src/renderer/views/preload/PreloadProgressView.tsx:178]: Uh is it okay though? I think we might be losing some tracking information of some crashes with unknown errors.

Comment (riemannulus) [reply src/renderer/views/preload/PreloadProgressView.tsx:178]: I think that's fine. There is only one case of going to the default where the Launcher get an unexpected error code from Headless, but it doesn't look good for the launcher to handle an undefined error code arbitrarily.

Comment (BasixKOR) [reply src/renderer/views/preload/PreloadProgressView.tsx:178]: Well if that's the case I don't see a problem too. Thanks for an explaination!

File: src/main/main.ts (hunk 1 above)
Comment (BasixKOR) [reply src/main/main.ts:574]: Did you forget to replace this one?

**Review by BasixKOR — APPROVED** [review]

**Review by moreal — APPROVED** [review]

**Pull request merged**